### PR TITLE
[gdal] Fix build for non-Windows targets

### PR DESCRIPTION
**Pull request by dg0yt**
- #### What does your PR fix?  
  This PR fixes building gdal for non-Windows systems (#9068).
  It also restructures the portfile for better maintainability.
  It includes significant patching to configure.ac. There seems to be no better way to handle vcpkg's release+debug filesystem layout.

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
  all, no

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
  yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
  yes

### CODE REVIEW

File: ports/gdal/vcpkg.json [hunk 1]
```diff
@@ -40,6 +42,10 @@
       "dependencies": [
         "libmariadb"
       ]
+    },
+    "system-libraries": {
+      "$supports": "!windows",
```
Comment (JackBoosY) [inline ports/gdal/vcpkg.json:47]:
```suggestion
      "supports": "!windows",
```
Should be `supports`?

Comment (dg0yt) [reply ports/gdal/vcpkg.json:47]: If "supports" was a supported field inside "features"... It is just documentation at the moment.

Comment (JackBoosY) [reply ports/gdal/vcpkg.json:47]: cc @strega-nil-ms

Comment (dg0yt) [reply ports/gdal/vcpkg.json:47]: @JackBoosY cf. https://github.com/microsoft/vcpkg/issues/17651

Comment (JackBoosY) [reply ports/gdal/vcpkg.json:47]: Fine, please run `x-add-version`.

**Comment by dg0yt** [comment]
Note to self: There are some conflicting gdal changes in https://github.com/microsoft/vcpkg/pull/15808 (draft) by @ras0219

**Comment by dg0yt** [comment]
Still waiting for approval of curl changes (#17790).
The pc file and the cmake wrapper need extra work.

**Comment by dg0yt** [comment]
TODO: gdal.pc

FTR, CI doesn't build many depending port at the moment:
- osx: none [!]
- linux: osg, osg-qt
- windows: osg, osg-qt, pdal, pdal-c
- x64_windows_static_md: osg, osg-qt, pdal

(And no gdal build for arm, uwp.)

**Comment by dg0yt** [comment]
This is almost complete.

The patch for using pkgconfig in connection with verifying configuration output seems to be a good test for pc file quality. minw-static has a lot of broken ones. (Fixing that issues is out of scope for this PR.)

Finally, I need to get the transitive dependencies into the cmake wrapper, for non-Windows and mingw (scope of this PR). Not all of them provide cmake configuration,so it is not easy to list them explicitly. Simplest thing might be querying pkgconfig, but still it takes some effort to deal with debug vs. release. At least, success can be verified with pdal.

**Comment by dg0yt** [comment]
> At least, success can be verified with pdal.

No, pdal is heavily patched and outdated. Can't work on it now. 

But I did test successfully with a minimal local cmake project on linux, with `GDALAllRegister()` in `main`. This PR is fine.

File: ports/gdal/vcpkg-cmake-wrapper.cmake [hunk 2]
```diff
@@ -1,12 +1,97 @@
...
+                break()
+            endif()
+        endforeach()
+        if(NOT TARGET ${dependency})
+            string(TOUPPER ${package} _gdal_deps_package)
+            if(DEFINED ${_gdal_deps_package}_LIBRARIES)
+                set(dependency ${${_gdal_deps_package}_LIBRARIES})
+            elseif(DEFINED ${package}_LIBRARIES)
+                set(dependency ${${package}_LIBRARIES})
+            elseif(DEFINED ${_gdal_deps_package}_LIBRARIES)
```
Comment (dg0yt) [inline ports/gdal/vcpkg-cmake-wrapper.cmake:49]: ..._LIBRARY

**Comment by JackBoosY** [comment]
Please ping me if this PR is ready for review.

**Comment by dg0yt** [comment]
Ping @JackBoos. GIthub states that I "marked this pull request as ready for review 2 days ago".

**Comment by dg0yt** [comment]
PS: I will fix the LIBRARY typo together with possible review comments.

**Comment by JackBoosY** [comment]
cc @Neumann-A for review this PR.

**Comment by dg0yt** [comment]
Kindly reminder for reviews. Cannot look at gdal updates, or upstreaming any changes, before sufficient agreement on the current PR. And a huge number of dependencies makes this PR volatile until merged.

**Comment by JackBoosY** [comment]
Can you please also fix https://github.com/microsoft/vcpkg/issues/17609?
See my [comment](https://github.com/microsoft/vcpkg/issues/17609#issuecomment-883114943).

**Comment by dg0yt** [comment]
> Can you please also fix #17609?

Hm, it is somewhat out of scope. This PR is mostly about the autotools build system, apart from general port file improvements.

#17609 is reported for Windows. The comments are using CMake language, but GDAL isn't. 

And I don't have the MSVC toolchain. I cannot build gdal[core,libspatialite]:x64-windows outside CI.

**Comment by JackBoosY** [comment]
@dg0yt  Okay, that's fine.

**Review by JackBoosY — APPROVED** [review]

**Comment by JackBoosY** [comment]
I'd like @Neumann-A review this PR.

**Comment by JackBoosY** [comment]
Does this issue also fix #15526?

**Comment by dg0yt** [comment]
After re-reading, #15526 is about using (static) gdal with MSVC. The whole area of *using gdal* will probably need more work. It depends on whether usage requirements are handled
- via pkgconfig files
- `FindGDAL.cmake` (certainly incomplete)
- manually (in MSVC projects?).

As I wrote in https://github.com/microsoft/vcpkg/pull/17698#issuecomment-873888821, there is a varying level of CI coverage for using gdal in different triplets. I guess there are also lots of workaround in consuming ports. Improving this situation is for coming PRs. The focus of this PR is on controlling what is pulled in when *building gdal*, thus stabilizing what needs to be pulled in when linking statically to gdal.

**Comment by mathisloge** [comment]
gdal 3.3.1 is released, too. https://github.com/OSGeo/gdal/releases/tag/v3.3.1

**Comment by mathisloge** [comment]
@dg0yt have tested gdal on windows with msvc (dynamic amd x64). 
This PR has also fixed some critical issues for me on windows. great work!

File: ports/gdal/portfile.cmake [hunk 3]
```diff
@@ -166,81 +172,197 @@ if (VCPKG_TARGET_IS_WINDOWS)
...
 
 else()
     # See https://github.com/microsoft/vcpkg/issues/16990
     vcpkg_execute_required_process(
-        COMMAND touch config.rpath
-        WORKING_DIRECTORY ${SOURCE_PATH}
+        COMMAND "${CMAKE_COMMAND}" -E touch config.rpath
+        WORKING_DIRECTORY "${SOURCE_PATH}"
         LOGNAME touch-${TARGET_TRIPLET}
     )
```
Comment (strega-nil-ms) [inline ports/gdal/portfile.cmake:191]:
```suggestion
        file(TOUCH "${SOURCE_PATH}/config.rpath")
```

**Review by strega-nil-ms — APPROVED** [review]
Just needs the minor change

File: ports/gdal/vcpkg-cmake-wrapper.cmake [hunk 4]
```diff
@@ -1,12 +1,97 @@
...
+    _gdal_add_dependency(netCDF::netcdf  netCDF CONFIG)
+    _gdal_add_dependency(openjp2  OpenJPEG CONFIG)
+    _gdal_add_dependency(PROJ::proj  PROJ4 CONFIG)
+    _gdal_add_dependency(unofficial::sqlite3::sqlite3  unofficial-sqlite3 CONFIG)
+    _gdal_add_dependency(TIFF::TIFF  TIFF)
+    _gdal_add_dependency(ZLIB::ZLIB  ZLIB)
+    _gdal_add_dependency(zstd::libzstd  zstd CONFIG)
+    if(";${ARGS};" MATCHES ";REQUIRED;" AND NOT GDAL_FOUND)
+        message(FATAL_ERROR "Failed to find dependencies of GDAL")
+    endif()
```
Comment (strega-nil-ms) [inline ports/gdal/vcpkg-cmake-wrapper.cmake:96]:
```suggestion
    if("REQUIRED" IN_LIST ARGS)
        message(FATAL_ERROR "Failed to find dependencies of GDAL")
    endif()
```

**Comment by dg0yt** [comment]
> gdal 3.3.1 is released, too. https://github.com/OSGeo/gdal/releases/tag/v3.3.1

I'm aware of that. But I can't push new things while waiting for review. This PR already triggered countless work on incoming dependencies.

File: ports/gdal/portfile.cmake (hunk 3 above)
Comment (dg0yt) [reply ports/gdal/portfile.cmake]: Oh no, this is the wrapper. `file(TOUCH ...)` is cmake 3.12. That's why I am asking about cmake versions!

Comment (dg0yt) [reply ports/gdal/portfile.cmake]: ... It is the port file. Sorry for the noise.

File: ports/gdal/vcpkg-cmake-wrapper.cmake (hunk 4 above)
Comment (dg0yt) [reply ports/gdal/vcpkg-cmake-wrapper.cmake]: But this is the wrapper. `IN_LIST` is cmake 3.4. I still wait for you to allow more than 3.1.

**Review by vicroms — APPROVED** [review]
LGTM waiting for CI results

File: ports/gdal/vcpkg-cmake-wrapper.cmake (hunk 4 above)
Comment (dg0yt) [reply ports/gdal/vcpkg-cmake-wrapper.cmake]: @strega-nil-ms We need the `AND NOT GDAL_FOUND` back. It is modified by the helper function if one of the dependencies is missing.

**Comment by dg0yt** [comment]
I verified that `find_package(GDAL REQUIRED)` broken now, due to the chances on the last mile, https://github.com/microsoft/vcpkg/pull/17698#discussion_r675349319. Next time I will revert to draft.

**Comment by mathisloge** [comment]
@dg0yt yep just runned into it at https://github.com/microsoft/vcpkg/pull/18849

**Pull request merged**